File: Data Structures & Algorithms/validate-parentheses/submission-0.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for char in s:
            if char in ('(', '{', '['): stack.append(char)
            else:
                if not stack: return False
                match char:
                    case ')':
                        if stack.pop() != '(':
                            return False
                    case '}':
                        if stack.pop() != '{':
                            return False
                    case ']':
                        if stack.pop() != '[':
                            return False
        return not stack

File: Data Structures & Algorithms/k-closest-points-to-origin/submission-0.cpp
class Solution {
public:
    constexpr static auto dist_cmp = [](const vector<int>& p1, const vector<int>& p2) constexpr {
        double dist1 = p1[0]*p1[0] + p1[1]*p1[1];
        double dist2 = p2[0]*p2[0] + p2[1]*p2[1];
        return dist1 < dist2;
    };
    
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        std::priority_queue<vector<int>, vector<vector<int>>, decltype(dist_cmp)> pq(points.begin(), points.begin()+k);

        for (int i = k; i < points.size(); ++i){
            pq.push(points[i]);
            pq.pop();
        }

        std::vector<vector<int>> results;
        while (!pq.empty()) {
            results.push_back(pq.top());
            pq.pop();
        }

        return results;

    }
};
